Subclass lazy loading

SanderBoelaars opened the pull request:
This PR is a continuation of the closed PR https://github.com/doctrine/orm/pull/10538 (which cannot be reopened, my apologies, I am new to making Pull Requests)

I have extracted the functionality for retrieving the discriminator column value into a separate function, like suggested by @beberlei.

I also noticed a grammatical error while reading https://github.com/doctrine/orm/blob/2.14.x/CONTRIBUTING.md and included a fix, I hope it is okay to do this in the same PR.

Commit:
Enabled lazy-loading for subclasses.

This enables lazy-loading of classes that have subclasses by using
the discriminator column to get the right subtype for the Proxy.

This also changes a few tests since lazy-loading subclasses
would now be possible and fetched classes are expected to be
an instance of the Proxy class.

Commit:
Enabled lazy-loading for subclasses.

This enables lazy-loading of classes that have subclasses by using
the discriminator column to get the right subtype for the Proxy.

This also changes a few tests since lazy-loading subclasses
would now be possible and fetched classes are expected to be
an instance of the Proxy class.

Commit:
Extracted code block into a new method.

Extracted the code block into a new method for
safety and maintaining more readability.

Commit:
Extracted code block into a new method.

Extracted the code block into a new method for
safety and maintaining more readability.

Commit:
Fixed a grammatical error.

Commit:
Fixed a grammatical error.

Commit:
Fixed coding standard errors.

Commit:
Fixed coding standard errors.

mpdude commented:
Maybe someone from the Doctrine team can reopen the old PR? There are some comments that give explanations and background, and it would be great if we could continue with these instead of starting with a blank slate here.

Mika56 committed:
docs: Remove incorrect @SequenceGenerator info (#10583)

The default allocationSize for @SequenceGenerator has actually been set to 1 ever since 434325e (back in 2010!)
This was done to remove confusion, but the docs were never updated to reflect the change

Commit:
Fixed more coding standard errors.

Commit:
Fixed more coding standard errors.

SanderBoelaars commented:
Can a maintainer please take a look at this: https://github.com/doctrine/orm/actions/runs/4436405563
It needs to be approved before it can start. 

Sadly I don't think it would be possible to reopen the closed PR's like @mpdude suggested because the relevant branches have been deleted.

e2palmes committed:
Added missing ';' 

missing ';' on Obtaining the EntityManager section, bootstrap code line 32

greg0ire committed:
Merge pull request #10589 from e2palmes/patch-1

greg0ire committed:
Upgrade to Psalm 5.9.0

It looks like some issues are now represented by different classes.

greg0ire committed:
Merge pull request #10604 from greg0ire/psalm-5.9.0

MatTheCat committed:
Handle null comparisons in `ManyToManyPersister` (#10587)

* Add test case for https://github.com/doctrine/orm/issues/7717

* Do not hide null equality checks in `SqlValueVisitor::walkComparison`

* Annotate `GH7717Parent::$children` type

ecourtial-vivlio committed:
fix typo in HydrationCompleteHandler doc

greg0ire committed:
Merge pull request #10620 from ecourtial/fix-doc-typo

fix typo in HydrationCompleteHandler doc

mever commented on lib/Doctrine/ORM/UnitOfWork.php:
loadOneToOneEntity performs a query using BasicEntityPersister::load(). And $this->getTargetSubClass() is also fetching data from the database. Both data can reasonably be assumed to be reached with an index. So, all in all, this seems like a very good feature.

mever reviewed (approved):
This feature looks very nice. It adorns Doctrine to include this PR and make the project yet another bit more awesome :)

greg0ire commented:
> Maybe someone from the Doctrine team can reopen the old PR? There are some comments that give explanations and background, and it would be great if we could continue with these instead of starting with a blank slate here.

For some reason I cannot.

tmihalicka committed:
Fixed xsd schema for support FQCN type

After update to orm 2.14.2 invalid xsd schema error is occured, when in field,id or attribute-override have type is FQCN

derrabus committed:
Run tests on SQLite with foreign keys enabled (#10632)

greg0ire committed:
Merge pull request #10630 from monadial/fix/fqcn-type-in-xml-mapping

Fixed xsd schema for support FQCN type

yobrx committed:
fix array association on partial index

greg0ire committed:
Merge pull request #10642 from yobrx/patch-1

Commit:
Fix single table inheritance with intermediate abstract class(es)

Fixes #10625

phansys committed:
Apply `SlevomatCodingStandard.Commenting.AnnotationName` CS rule (#10653)

greg0ire committed:
Remove JoinColumn from inverse side (#10652)

This makes no sense because it describes a one-to-one where each table
references the other one. I do not think this is deliberately supported,
and it probably will not be supported at all in 3.0.

derrabus committed:
PHPStan 1.10.14 (#10655)

mever commented:
@beberlei , can you spare some time to approve/disapprove this pull request? We would like to upgrade to a version of Doctrine that has this feature.

mpdude commented:
As I've written in https://github.com/doctrine/orm/pull/10538#issuecomment-1445337437, to me it seems a bit strange that the `UnitOfWork` until now does not do any SQL querying/processing. So to me it seems your new code is in the wrong place, if it all of a sudden makes the UoW build SQL queries.

Imagine we'd need some more abstraction, like for SQL variants. Where would you place this code then? Can you find similar SQL code there already, which would be an indication of being in the right place architecture-wise?

mever commented:
Hi @mpdude , first I was thinking about placing it in the EntityPersister because it executes queries. But then I realized it does not deal with proxies, which is exactly what we want. That said, maybe the ProxyFactory would be a better place.

One method there looks like a good candidate: ProxyFactory::getProxy(), it can return a subclass of the first parameter and still be compliant with Liskov Substitutability. The downside of that approach, however, is that then the proxy factory is going to execute queries. Or at least for proxies that do not have a fixed type (i.e.: in the case of subclasses that aren't proxies themselves). To communicate intent, we could also add ProxyFactory::getSubclassProxy() which only deals with inherited entities but is functionally equivalent to ProxyFactory::getProxy(). That would be more transparent, and adding a comment about it executing a query to fetch subclass information from the database would be really clear for people that care about performance, as we do ;)

What do you think about splitting the methods or should we combine them to increase abstraction? Or were you thinking in a different direction?

SanderBoelaars commented:
Hello @gregOire @mpdude @beberlei, Like @mever, I myself was also looking at placing the code in the EntityPersister, but after reading @mever's message I do agree that ProxyFactory is probably a better class for this. Like @beberlei suggested earlier (https://github.com/doctrine/orm/pull/10538) I have extracted the code which executes the extra query into a separate function and I have now moved this function to the ProxyFactory. Would this be acceptable? If not, maybe EntityPersister would be a good place after all?

greg0ire committed:
Merge pull request #10643 from htto/bugfix/sti-with-abstract-intermediate

Fix single table inheritance with intermediate abstract class(es)

mever commented:
@beberlei What do you think? Should we put some effort into this in order to close this PR?

mpdude commented:
The test case you added is valueable: I used it to step through the code and to me it seems `\Doctrine\ORM\Persisters\Entity\BasicEntityPersister::getSelectColumnsSQL` would be a good place for (some) of these changes.

That method already contains some checks to see if we're associated with other entities, to consider the type of association and if inheritance is used. I haven't thought through all the possible combinations/association types, but to me it seems this method rather `continue`s early in these challenging situations.

What you could do there is to JOIN the table of the referred-to entity. But, instead of adding all the fields and potential associations, select just the discriminator column. Most of the code should already  be there.

Also, make sure to add the relevant information to the ResultSetMapping, but make sure to call `addMetaResult` only (indicating names/aliases used for information like a discriminator column).

By looking at `BasicEntityPersister::getSelectColumnsSQL()`, you also see things you potentially need to deal with when building SQL selects yourself – namely the `quoteStrategy` or the `filterSql` mechanism.

Having the changes in `BasicEntityPersister` puts them in a class that is already involved in building SQL queries, instead of adding this as a new concern to the `UnitOfWork`.

Also, it should be possible to load at least the discriminator values _for all rows at once_ instead of having to look them up one by one for every entity being hydrated.

Commit:
Made subclass proxies possible without separate query.

Changed functionality to no longed execute a separate query to retrieve
the discriminator column of object associations for creation of Proxies.

Commit:
Merge branch 'subclass-lazy-loading' of github.com:SanderBoelaars/orm into subclass-lazy-loading

# Conflicts:
#	CONTRIBUTING.md
#	lib/Doctrine/ORM/UnitOfWork.php
#	tests/Doctrine/Tests/ORM/Functional/ClassTableInheritanceTest.php
#	tests/Doctrine/Tests/ORM/Functional/Ticket/GH10466Test.php

mever commented:
Hi @mpdude , can you take a look at this? It would be nice if this can be default feature of Doctrine ORM.
